dot/sync/service.go

```diff
@@ -125,6 +127,13 @@ func NewSyncService(cfgs ...ServiceConfig) *SyncService {
 		cfg(svc)
 	}
 
+	// Set initial strategy
+	if svc.warpSyncStrategy != nil {
+		svc.currentStrategy = svc.warpSyncStrategy
+	} else {
+		svc.currentStrategy = svc.fullSyncStrategy
+	}
+
```

Comment:
Can we make it a `ServiceConfig`? `WithCurrentStrategy(s Strategy) { svc.currentStrategy = s } `

Reply:
I can add that option too, but I think this is something the service should decide not the caller

Reply:
I'm thinking that could be better to get rid of the `WithWarpSyncStrategy()` and `WithFullSyncStrategy` and let the sync service create this strategies by itself instead of relying on the caller.

Reply:
The downside of doing that is if some strategies have dependencies that the service doesn't need to know about, they could be passed directly to the strategy when instantiated, instead of passing them to the service. Not sure if that's currently the case or likely to be the case in the future though.

And we'd need another method of telling the service which strategy to use initially.

Reply:
I think the strategy initialization and the strategy selection is part of the sync service internal implementation, the caller shouldn't know about it to prevent coupling our solution.

Reply:
I agree with Haiko, I don't really see a requirement to move away from the `WithStrategies(current, default)` API.  Seems to be quite strategy specific now by introducing `SyncService.fullSyncStrategy` and `SyncService.warpSyncStrategy` into the `SyncService` state.  Given that the initial `WithStrategies` call supported a current and default strategy, I would've assumed if we were to introduce a helper `WithWarpSync`method, it would just call `WithStrategies` with the actual strategies doing all the provider stuff within the method, instead of taking in the parameter of the strategy.

Reply:
I strongly believe that keeping the strategy selection and initialization within the sync service’s internal implementation is the better approach, and here’s why:

- Better encapsulation: by having the sync service handle its own strategy setup and selection, we avoid coupling the selection logic to external callers. This allows the service to evolve internally without requiring changes from every caller whenever we add or adjust a strategy. While this might mean we need to provide additional dependencies when creating the sync service, that’s due to the lack of a more comprehensive dependency injection mechanism.

- Avoiding exposure of internal implementations: strategy selection should be an implementation detail managed by the sync service, not something callers need to be aware of or control. Callers should only need to interact with the service without knowing the internal details of how it chooses or initializes the strategy—or even that we’re using "strategies" at all.

- Open to change: If we decide to include more than two strategies (like the `stateSyncStrategy` I'm currently working on), or if we restructure how strategies are used, having the sync service manage its own strategy logic will make these changes significantly easier. It allows us to avoid adjustments in each external caller and its tests, keeping the interface simpler and more resilient to internal changes.

In short, managing the strategies internally not only keeps the design cleaner but also future-proofs it.

Apologies if I sound overly firm, but I genuinely believe this is the right path forward. That said, if after considering these points you still feel this approach is flawed, I'm happy to go along with the current implementation 😄

Reply:
I guess I don't really see the reason to change much of this existing code.  I attempted to try and see how it would have looked without these changes (see 5b0207b190d6b165fd6675c03bda93b537e5ce5e), and it seems inline with the approach that @EclesioMeloJunior authored. 

I would give @EclesioMeloJunior final authority here.  He already approved this PR, but I would like to hear what he has to say on this thread.

Reply:
> I think the strategy initialization and the strategy selection is part of the sync service internal implementation, the caller shouldn't know about it to prevent coupling our solution.

That's not true for polkadot-sdk, basically the caller can decide what strategy the sync should use, that decision is also made via cli `--warp-sync`. 

So, I would say that we should stick with `currentStrategy` and `defaultStrategy` and enable the definitions of both to the caller.

With regards the points @dimartiro put:
- Better encapsulation: I disagree on that since we should enable the caller/user choose the desired sync strategy, that is easy due to the usage of interfaces which enables a less encapsulation (we don't depend on the concrete impl, since we can just provide whatever satisfies the `Strategy` interface) 
- Avoiding exposure of internal implementations: Here is the point, we should not expose the implementation to the part that is using it, which means that the sync engine should have exposure just for the interface methods. Each strategy should be self contained, the caller is the one who provides us one valid implementation, you can imagine sync as a modular system where you can plug and un-plug strategies at runtime.
- Open to change: If we hardcode that warp sync is the default strategy then we will might face problems to sync if we start from scratch or if we start a single node, what I am suggesting is make it as less coupled as we can where we leave the choice to the caller/user, so depending on its needs.

Reply:
@dimartiro makes a good point regarding the state sync strategy. I know we discussed this before, but I don't remember how we decided to handle this. IIRC, state sync needs to run concurrently to full sync after warp sync has completed. In our previous discussion I think we mentioned two options:

1. Drive state sync from `FullSyncStrategy`

For this `FullSyncStrategy` needs to be informed that it also needs to run state sync, in the case where warp sync ran to completion before `FullSyncStrategy` is started. I assume @dimartiro intends to use the [added `Result()` method](https://github.com/ChainSafe/gossamer/pull/4275#discussion_r1852981665) in the `Strategy` interface for this somehow.

2. Drive state sync from `SyncService`

For this, the service needs to figure out when it needs to run state sync strategy in addition to `FullSyncStrategy`. That would require keeping more than two strategies in the service and changing the API discussed in this thread as well.

As I mentioned before, I don't think it matters that much since I don't see us adding many new sync strategies in the future.

Reply:
@EclesioMeloJunior Regarding this:

> That's not true for polkadot-sdk, basically the caller can decide what strategy the sync should use, that decision is also made via CLI --warp-sync.

We need to differentiate between two aspects here. On one hand, we have the `sync mode`, which could be either `warp` or `full` sync. On the other hand, we have the implementation, i.e., the strategies.  
In my opinion, the caller shouldn't need to know about the strategies but only about the `sync mode`. The sync mode is a parameter required by the sync service, but the internal implementation, including decisions about where and when to configure, instantiate, and use strategies, should be the responsibility of the sync service, not the caller.  

For instance, suppose we decide to change the approach and stop using strategies in favor of another solution. Even if we currently have an interface in place, we would still need to modify the caller to adapt to the new approach. My main point is to avoid this by encapsulating all the sync logic within the sync service.

Regarding what Substrate does: the sync service is responsible for creating the strategies, not the caller. You can take a look [here](https://github.com/paritytech/polkadot-sdk/blob/6d5f8141ad42d7a528a09715c492974550709b92/substrate/client/network/sync/src/strategy/polkadot.rs#L355-L391).

As @haikoschol pointed out, we are going to need to use "state sync" in the future. Are we expecting the caller to configure, instantiate, and send the 3 strategies to the sync service instead of letting the sync service decide which strategy to instantiate and use?  

In my opinion, that's not the right architectural decision for our system. However, I’m open to accepting your approach and leaving it as is if you prefer. I just wanted to raise these concerns beforehand.

Reply:
@haikoschol `state sync` should not run in parallel with `full sync` after `warp sync finishes`, what happens is that `warp sync` finishes and then starts `state sync` that should be completed before starting `full sync`. What should happens "together" with full sync is the `gap sync` which is this case will be part of full sync and not a strategy under its own.

@dimartiro yeah, I saw the changes and maybe then we can do is:

```go
// Set initial strategy
func WithWarpSyncStrategy(warpSyncStrategy Strategy) ServiceConfig {
	return func(svc *SyncService) {
		svc.warpSyncStrategy = warpSyncStrategy
                 svc.currentStrategy = warpSyncStrategy
	}
}

func WithFullSyncStrategy(fullSyncStrategy Strategy) ServiceConfig {
	return func(svc *SyncService) {
		svc.fullSyncStrategy = fullSyncStrategy
                svc.currentStrategy = fullSyncStrategy
	}
}
```

the you don't need the if/else, since when the strategy is defined by the config you already set it as current.

Reply:
@EclesioMeloJunior thanks for clarifying that!

IMHO your proposed changes make the behavior very confusing. Looking at the two function signatures, I would never expect that the order in which I call them matters or that they overwrite `svc.currentStrategy`.

Reply:
Let's proceed as follows: since this discussion is taking too many iterations and blocking this PR, I'd prefer to use the simpler version with `current` and `default` for now. However, I believe we'll need to revisit this discussion in the next PR when I introduce the state sync.